### DISCUSSION TITLE
Javascript of third-party library not found

### DISCUSSION THREAD

**Discussion by dmenne** [Appsilon/rhino#279] [Q&A] [answered]
There seems to be a problem with some external Javascript libraries. In the example below, I get browser console error messages complaining about shinyBS.css and shinyBS.js. I know that shinyBS is no longer maintained, and the author is unresponsive, but it still works with standard Shiny apps. 

When I uncomment `library(shinyBS)` the error disappears, but that's a no-go and gives a box warning.

```
box::use(
  shiny[actionButton, fluidPage, moduleServer, NS],
  shinyBS[bsTooltip]
)

#library(shinyBS)


#' @export
ui <- function(id) {
  ns <- NS(id)
  fluidPage(
    actionButton(ns("hello"), "Hello"),
    bsTooltip(ns("hello"),
               "Pressing the button does nothing",
              options = list(container = "body")
    )
  )
}

#' @export
server <- function(id) {
  moduleServer(id, function(input, output, session) {
  })
}

**Comment by dmenne** [comment, marked as answer]
A workaround for packages that fail to add the correct resource path. The example is for `shinyBS`

- add `library(shinyBS)`; don't care about the `module may cause issues` warning
- When things work with this addition:
- `print(resourcePaths())` after `library shinyBS`
- You will see a list, and the relevant line is 
``` 
                            sbs 
                          "C:/R/R-4.2.0/library/shinyBS/www" 
```
- remove the library call, add
```
shiny::addResourcePath("sbs", system.file("www", package = "shinyBS"))
```

No more warnings

**Comment by kamilzyla** [comment]
One of my colleagues faced similar problems. This is my answer to his questions - perhaps it can be useful here too. There are two issues with shinyBS; see below for explanations and workarounds which can be used in Rhino.

### Resource path setup

shinyBS uses the `.onAttach()` hook to call `shiny::addResourcePath()` which is necessary for its resources (JS and CSS) to be loaded correctly. This hook is never run with `box::use()` (normally a `library()` call would run it). It is worth noting that the problem will be also present in vanilla Shiny if you only use `::` to access shinyBS functions (shinyBS should use `.onLoad()` instead).

**Workaround:** Add the following snippet to your `app/main.R` (just below the `box::use()` statements is a good place):

```r
# Run the `.onAttach` hook (shinyBS uses it to add a Shiny resource path).
suppressWarnings(library(shinyBS))
```

You'll still need to explicitly `box::use()` whatever functions you need from  shinyBS (`library()` doesn't work in a box module).

### IDs in Shiny modules

shinyBS doesn't automatically apply `session$ns` to the IDs in the server of a Shiny module (unlike e.g. shinyjs).

**Workaround:** Apply `session$ns` manually, e.g.:

```r
#' @export
ui <- function(id) {
  ns <- NS(id)
  bootstrapPage(
    bsAlert(ns("alert"))
  )
}

#' @export
server <- function(id) {
  moduleServer(id, function(input, output, session) {
    createAlert(
      session = session,
      anchorId = session$ns("alert"),
      title = "Title",
      content = "Content."
    )
  })
}
```
